### PR TITLE
#213 Add command to run a new CI Build (Continuation of PR #224)

### PR DESCRIPTION
**Pull request by lufty-demansol**
This PR is forked of [#224](https://github.com/mattermost/mattermost-plugin-gitlab/pull/224) to continue the work.

### CODE REVIEW

File: server/gitlab/api.go [hunk 1]
```diff
@@ -412,3 +412,23 @@ func (g *gitlab) ResolveNamespaceAndProject(
...
+func (g *gitlab) TriggerNewBuildPipeline(user *UserInfo, repo string, refToCommit string) (*TriggerPipelineResponse, error) {
+	client, err := g.gitlabConnect(*user.Token)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, ctxCancel := context.WithTimeout(context.Background(), DefaultRequestTimeout)
+	defer ctxCancel()
+
+	pipeline, response, gitError := client.Pipelines.CreatePipeline(repo, &internGitlab.CreatePipelineOptions{
```
Comment (hanzei) [inline server/gitlab/api.go:425]: Why not `err`?

File: server/command.go [hunk 2]
```diff
@@ -238,6 +239,23 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
 		response := p.getCommandResponse(args, message)
 		return response, nil
 
+	case commandBuild:
+		if len(parameters) != 2 {
+			return p.getCommandResponse(args, "Command arguments mismatched, please use `/gitlab help` to see action parameters. "), nil
+		}
+		project := parameters[0]
+		ref := parameters[1]
```
Comment (hanzei) [inline server/command.go:247]: Should the command  default to default branch?

File: server/gitlab/user.go [hunk 3]
```diff
@@ -24,6 +24,11 @@ type UserSettings struct {
 	Notifications  bool   `json:"notifications"`
 }
 
+type CommitDetails struct {
+	Repository string `json:"repo_name_or_id"`
+	Branch     string `json:"reference"`
+}
+
```
Comment (hanzei) [inline server/gitlab/user.go:31]: Is this used?

File: server/gitlab/api.go (hunk 1 above)
Comment (maisnamrajusingh) [reply server/gitlab/api.go]: My bad.

File: server/command.go (hunk 2 above)
Comment (lufty-demansol) [reply server/command.go:247]: my bad

File: server/gitlab/api.go (hunk 1 above)
Comment (lufty-demansol) [reply server/gitlab/api.go]: my bad. I have updated it.

File: server/gitlab/user.go (hunk 3 above)
Comment (maisnamrajusingh) [reply server/gitlab/user.go]: this is removed now.

**Comment by maisnamrajusingh** [comment]
@hanzei  @iomodo can you give approval to @lufty-demansol  as he is not able to run the CI here.

File: .gitignore [hunk 4]
```diff
@@ -1 +1,3 @@
 dist
+
+.idea/
```
Comment (hanzei) [inline .gitignore:3]: 2/5 that editor configs should be in the global `gitignore `

File: go.mod [hunk 5]
```diff
@@ -4,12 +4,13 @@ go 1.13
 
 require (
 	github.com/golang/mock v1.6.0
-	github.com/mattermost/mattermost-plugin-api v0.0.21
-	github.com/mattermost/mattermost-server/v6 v6.0.3
-	github.com/mholt/archiver/v3 v3.5.0
+	github.com/mattermost/mattermost-plugin-api v0.0.12-0.20200908143138-66edf222f7ea
+	github.com/mattermost/mattermost-server/v5 v5.25.1 // indirect
+	github.com/mattermost/mattermost-server/v6 v6.4.2
```
Comment (hanzei) [inline go.mod:9]: Why this change?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: server/command.go (hunk 2 above)
Comment (lufty-demansol) [reply server/command.go]: fixed by https://github.com/mattermost/mattermost-plugin-gitlab/pull/280/commits/796f2286fe8580ec4dc31f98aef9de01e745f3aa

File: .gitignore (hunk 4 above)
Comment (lufty-demansol) [reply .gitignore]: Fixed by https://github.com/mattermost/mattermost-plugin-gitlab/pull/280/commits/aea509ab9f6b1b9483a24114a693f64c2306ca7c

File: go.mod (hunk 5 above)
Comment (lufty-demansol) [reply go.mod]: Reverted back to v5: https://github.com/mattermost/mattermost-plugin-gitlab/pull/280/commits/ad4533fb4bcd82272aed341bd922482382e2386a

Comment (lufty-demansol) [reply go.mod]: I think it is already using v6 .. this comment is from master branch mentioning v6: https://github.com/mattermost/mattermost-plugin-gitlab/blob/master/server/plugin.go#L305

**Comment by lufty-demansol** [comment]
there is panic message that I don't quite understand yet
```
goroutine 19 [select]:
net/http.(*Transport).getConn(0xc0003fe780, 0xc0003fcb80, {{}, 0x0, {0xc0001c2900, 0x5}, {0xc000410812, 0xe}, 0x0})
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/transport.go:1372 +0xa45
net/http.(*Transport).roundTrip(0xc0003fe780, 0xc0001f8800)
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/transport.go:581 +0xcdf
net/http.(*Transport).RoundTrip(0xc0001f8800, 0x1692c00)
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/roundtrip.go:18 +0x37
net/http.send(0xc0001f8800, {0x1692c00, 0xc0003fe780}, {0x0, 0x108ff20, 0x0})
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:252 +0x93f
net/http.(*Client).send(0xc000406ba0, 0xc0001f8800, {0x12a75ce, 0x14, 0x0})
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:176 +0x158
net/http.(*Client).do(0xc000406ba0, 0xc0001f8800)
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:725 +0x1087
net/http.(*Client).Do(...)
	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:593
github.com/xanzy/go-gitlab.(*Client).configureLimiter(0xc000240800, {0x169b4c8, 0xc000237380})
	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/gitlab.go:479 +0x185
github.com/xanzy/go-gitlab.(*Client).Do.func1()
	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/gitlab.go:737 +0xc5
sync.(*Once).doSlow(0xc000240814, 0xc00020ba50)
	/usr/local/Cellar/go/1.17.5/libexec/src/sync/once.go:68 +0x128
sync.(*Once).Do(0xc000240814, 0x0)
	/usr/local/Cellar/go/1.17.5/libexec/src/sync/once.go:59 +0x47
github.com/xanzy/go-gitlab.(*Client).Do(0xc000240800, 0xc0003d6790, {0x1546d20, 0xc00027e380})
	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/gitlab.go:737 +0xc5
github.com/xanzy/go-gitlab.(*PipelinesService).CreatePipeline(0xc00018a6d8, {0x1537d20, 0xc0003d6560}, 0x2, {0xc00019bcb8, 0x1, 0x1})
	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/pipelines.go:290 +0x1a6
github.com/mattermost/mattermost-plugin-gitlab/server/gitlab.(*gitlab).TriggerNewBuildPipeline(0x189dbd0, 0xc000050de8, {0x15cedd9, 0x8}, {0x15cbf38, 0x6})
	/Users/arknable/Documents/Project/lufty-demansol/mattermost-plugin-gitlab/server/gitlab/api.go:427 +0x3c5
github.com/mattermost/mattermost-plugin-gitlab/server/gitlab.TestTriggerNewBuildPipeline(0xc000276820)
	/Users/arknable/Documents/Project/lufty-demansol/mattermost-plugin-gitlab/server/gitlab/api_test.go:23 +0x125
testing.tRunner(0xc000276820, 0x1619e08)
	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1306 +0x727
```

**Comment by lufty-demansol** [comment]
> there is panic message that I don't quite understand yet when running `make test`
> 
> ```
> goroutine 19 [select]:
> net/http.(*Transport).getConn(0xc0003fe780, 0xc0003fcb80, {{}, 0x0, {0xc0001c2900, 0x5}, {0xc000410812, 0xe}, 0x0})
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/transport.go:1372 +0xa45
> net/http.(*Transport).roundTrip(0xc0003fe780, 0xc0001f8800)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/transport.go:581 +0xcdf
> net/http.(*Transport).RoundTrip(0xc0001f8800, 0x1692c00)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/roundtrip.go:18 +0x37
> net/http.send(0xc0001f8800, {0x1692c00, 0xc0003fe780}, {0x0, 0x108ff20, 0x0})
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:252 +0x93f
> net/http.(*Client).send(0xc000406ba0, 0xc0001f8800, {0x12a75ce, 0x14, 0x0})
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:176 +0x158
> net/http.(*Client).do(0xc000406ba0, 0xc0001f8800)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:725 +0x1087
> net/http.(*Client).Do(...)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/net/http/client.go:593
> github.com/xanzy/go-gitlab.(*Client).configureLimiter(0xc000240800, {0x169b4c8, 0xc000237380})
> 	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/gitlab.go:479 +0x185
> github.com/xanzy/go-gitlab.(*Client).Do.func1()
> 	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/gitlab.go:737 +0xc5
> sync.(*Once).doSlow(0xc000240814, 0xc00020ba50)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/sync/once.go:68 +0x128
> sync.(*Once).Do(0xc000240814, 0x0)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/sync/once.go:59 +0x47
> github.com/xanzy/go-gitlab.(*Client).Do(0xc000240800, 0xc0003d6790, {0x1546d20, 0xc00027e380})
> 	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/gitlab.go:737 +0xc5
> github.com/xanzy/go-gitlab.(*PipelinesService).CreatePipeline(0xc00018a6d8, {0x1537d20, 0xc0003d6560}, 0x2, {0xc00019bcb8, 0x1, 0x1})
> 	/Users/arknable/Documents/GOPATH/pkg/mod/github.com/xanzy/go-gitlab@v0.60.0/pipelines.go:290 +0x1a6
> github.com/mattermost/mattermost-plugin-gitlab/server/gitlab.(*gitlab).TriggerNewBuildPipeline(0x189dbd0, 0xc000050de8, {0x15cedd9, 0x8}, {0x15cbf38, 0x6})
> 	/Users/arknable/Documents/Project/lufty-demansol/mattermost-plugin-gitlab/server/gitlab/api.go:427 +0x3c5
> github.com/mattermost/mattermost-plugin-gitlab/server/gitlab.TestTriggerNewBuildPipeline(0xc000276820)
> 	/Users/arknable/Documents/Project/lufty-demansol/mattermost-plugin-gitlab/server/gitlab/api_test.go:23 +0x125
> testing.tRunner(0xc000276820, 0x1619e08)
> 	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259 +0x230
> created by testing.(*T).Run
> 	/usr/local/Cellar/go/1.17.5/libexec/src/testing/testing.go:1306 +0x727
> ```

well CI passed so I guess it's not meant to be executed in local

**Review by hanzei — COMMENTED** [review]
The diff is wild. Please double check the changes.

**Review by hanzei — COMMENTED** [review]
The diff is wild. Please double check the changes.

**Comment by lufty-demansol** [comment]
@hanzei  please take a look

File: plugin.json [hunk 6]
```diff
@@ -91,4 +91,4 @@
             }
         ]
     }
-}
```
Comment (maisnamrajusingh) [inline plugin.json:94]: @dimashasbi this should be the same as master. Can you revert this as this is not a related change

Comment (lufty-demansol) [reply plugin.json:94]: on this commit [2f6fdcad](2f6fdcad4dace4a27e7e59fe46c58729d9019945) is already equal to master

File: server/gitlab/api.go [hunk 7]
```diff
@@ -412,3 +412,23 @@ func (g *gitlab) ResolveNamespaceAndProject(
 	}
 	return "", "", ErrNotFound
 }
+
+func (g *gitlab) TriggerNewBuildPipeline(user *UserInfo, repo string, refToCommit string) (*TriggerPipelineResponse, error) {
```
Comment (mickmister) [inline server/gitlab/api.go:416]: Can this function just return `go-gitlab`'s `Pipeline` struct instead of converting to our own type? Then there is no need for `server/gitlab/types.go`

File: server/gitlab/types.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	Ref         string     `json:"ref"`
+	Tag         bool       `json:"tag"`
+	Username    string     `json:"username"`
+	Name        string     `json:"name"`
+	UpdatedAt   *time.Time `json:"updated_at"`
+	CommittedAt *time.Time `json:"committed_at"`
+	WebURL      string     `json:"web_url"`
+}
+
+func newTriggerPipelineResponse(pipeline *internGitlab.Pipeline) *TriggerPipelineResponse {
```
Comment (mickmister) [inline server/gitlab/types.go:21]: I think `internGitlab` is supposed to be the `server/gitlab` package and _not_ the `go-gitlab` package. Also, see the previous comment about removing this file

File: server/webhook/issue.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ func (w *webhook) handleDMIssue(event *gitlab.IssueEvent) ([]*HandleWebhook, err
 
 	switch event.ObjectAttributes.Action {
 	case actionOpen:
-		if len(event.Assignees) > 0 {
+		if event.Assignees != nil && len(event.Assignees) > 0 {
```
Comment (mickmister) [inline server/webhook/issue.go:29]: Go automatically checks if this is `nil` before checking its length, so this change is unnecessary

Comment (mickmister) [reply server/webhook/issue.go:29]: Also, the changes in this file seem irrelevant to this PR. Please raise these changes in another issue/PR

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work @lufty-demansol! I have a few questions and requests

**Comment by hanzei** [comment]
@lufty-demansol Heads up that you need to merge `master`

File: plugin.json (hunk 6 above)
Comment (lufty-demansol) [reply plugin.json]: fixed here 85c440123cdac0b74a6ee6edcb6ef8ecc80ba37b

File: .gitignore [hunk 10]
```diff
@@ -1 +1 @@
-dist
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply .gitignore]: fixed here 0b9ec0130a6eda7c1ad81183d02a446f0399dae5

File: server/gitlab/api.go (hunk 7 above)
Comment (lufty-demansol) [reply server/gitlab/api.go]: fixed here 79db715b51441c46a443ddc9387452f1dd08b362

File: server/webhook/issue.go (hunk 9 above)
Comment (lufty-demansol) [reply server/webhook/issue.go]: fixed here 4146d4ba7739302a496d23d0017c2e0f4ca30732

File: server/webhook/merge_request.go [hunk 11]
```diff
@@ -96,7 +96,16 @@ func (w *webhook) handleChannelMergeRequest(event *gitlab.MergeEvent) ([]*Handle
...
+			label := sub.Label()
+			hasLabel := false
+			for _, l := range event.Labels {
+				if l.Name == label {
+					hasLabel = true
+					break
+				}
+			}
+
+			if sub.Label() != "" && !hasLabel {
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply server/webhook/merge_request.go]: yeah, no need to change. 
fixed here 624e346cb9ac2ec0deed0c19af3669d27f5a7569

File: server/command.go [hunk 12]
```diff
@@ -238,6 +239,28 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
...
 
+	case commandBuild:
+		lenParams := len(parameters)
+		if lenParams < 1 {
+			return p.getCommandResponse(args, "Command arguments mismatched, please use `/gitlab help` to see action parameters. "), nil
+		}
+		project := parameters[0]
+
+		ref := "main"
+		if lenParams == 2 {
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply server/command.go:290]: yeah, 
so, I'm providing validation and direct instruction on this commit ade0eaeed55c2d2f50584016a8ee7f711ce2afe9

File: server/command.go [hunk 13]
```diff
@@ -238,6 +239,28 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
...
+			ref = parameters[1]
+		}
+
+		pipeline, err := p.GitlabClient.TriggerNewBuildPipeline(info, project, ref)
+		if err != nil {
+			if strings.Contains(err.Error(), projectNotFoundError) {
+				return p.getCommandResponse(args, projectNotFoundMessage+project), nil
+			}
+			return p.getCommandResponse(args, "Encountered an error while triggering your pipeline"), nil
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply server/command.go]: Maybe, we should provide all the information about the error from the gitlab server.
On this commit 3c9806a7c0361d197afb547694909080392f6caf, I'm passing an error message to send to the front

**Comment by lufty-demansol** [comment]
> @lufty-demansol Heads up that you need to merge `master`

@hanzei , I have merged with master on this commit 9bf44b35410603bc5c6a302fef17040a41a8c265

**Comment by lufty-demansol** [comment]
@mickmister , I have answered all questions. please take a look.
